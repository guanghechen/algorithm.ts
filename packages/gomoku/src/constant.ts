import type { IShapeCount } from './types'

export enum GomokuDirectionType {
  LEFT = 0,
  TOP_LEFT = 2,
  TOP = 4,
  TOP_RIGHT = 6,
  RIGHT = 1,
  BOTTOM_RIGHT = 3,
  BOTTOM = 5,
  BOTTOM_LEFT = 7,
}

// Canvas x-arias direction.
// Directions in canvas coordinates.
export const gomokuDirections: Array<[dr: number, dc: number]> = Array.from(
  Object.entries({
    [GomokuDirectionType.TOP]: [-1, 0],
    [GomokuDirectionType.TOP_RIGHT]: [-1, 1],
    [GomokuDirectionType.RIGHT]: [0, 1],
    [GomokuDirectionType.BOTTOM_RIGHT]: [1, 1],
    [GomokuDirectionType.BOTTOM]: [1, 0],
    [GomokuDirectionType.BOTTOM_LEFT]: [1, -1],
    [GomokuDirectionType.LEFT]: [0, -1],
    [GomokuDirectionType.TOP_LEFT]: [-1, -1],
  }).reduce((acc, [key, value]) => {
    // eslint-disable-next-line no-param-reassign
    acc[key] = value
    return acc
  }, []),
)

export const leftHalfGomokuDirectionTypes: GomokuDirectionType[] = [
  GomokuDirectionType.LEFT,
  GomokuDirectionType.TOP_LEFT,
  GomokuDirectionType.TOP,
  GomokuDirectionType.TOP_RIGHT,
]

export const rightHalfGomokuDirectionTypes: GomokuDirectionType[] = [
  GomokuDirectionType.RIGHT,
  GomokuDirectionType.BOTTOM_RIGHT,
  GomokuDirectionType.BOTTOM,
  GomokuDirectionType.BOTTOM_LEFT,
]

export const gomokuDirectionTypes = leftHalfGomokuDirectionTypes.concat(
  rightHalfGomokuDirectionTypes,
)

export const continuouslyShapeScoreMap: IShapeCount[] = [
  [0, 0, 0], // 0
  [0, 1, 8], // 1
  [0, 2, 16], // 2
  [0, 32, 64], // 3
  [0, 520, 520], // 4
]
