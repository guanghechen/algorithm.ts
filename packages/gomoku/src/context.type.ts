import type { GomokuDirectionType } from './constant'
import type { IGomokuBoard, IGomokuPiece, INeighbor } from './types'

export interface IGomokuContext {
  /**
   * Number of rows.
   */
  readonly MAX_ROW: number

  /**
   * Number of columns.
   */
  readonly MAX_COL: number

  /**
   * The maximum number of pieces with the same color allowed to be adjacent in the same direction.
   */
  readonly MAX_ADJACENT: number

  /**
   * The distance to the farthest neighbor to the current position.
   */
  readonly MAX_DISTANCE_OF_NEIGHBOR: number

  /**
   * The total number of valid positions on the board.
   */
  readonly TOTAL_POS: number

  /**
   * Gomoku board.
   */
  readonly board: Readonly<IGomokuBoard>

  /**
   * The total number of positions that haven been placed.
   */
  readonly placedCount: number

  /**
   * The coordinate id of the position in the middle of the chessboard.
   */
  readonly MIDDLE_POS: number

  /**
   * Initialize context with given pieces.
   * @param pieces
   */
  init(pieces: ReadonlyArray<IGomokuPiece>): void

  /**
   * Place a piece on the given position.
   * @param posId
   * @param playerId
   */
  forward(posId: number, playerId: number): boolean

  /**
   * Remove the piece from the given position.
   * @param posId
   */
  revert(posId: number): boolean

  /**
   * Get position id.
   * @param r Row number of the target position.
   * @param c Column number of the target position.
   */
  idx(r: number, c: number): number

  /**
   * Get the coordinates (r, c) of the target position.
   * @param posId
   */
  revIdx(posId: number): Readonly<[r: number, c: number]>

  /**
   * Check if a given position is a legal position
   * @param r
   * @param c
   */
  isValidPos(r: number, c: number): boolean

  /**
   * Move the specified number of steps in the given direction.
   *
   * If the position after the move is out of bounds, return -1.
   *
   * @param posId     Started position. !!!Should be a valid pos id in bounds.
   * @param dirType   Moving direction.
   * @param step      Number of steps to move. !!! Should be a non-negative integer.
   */
  safeMove(posId: number, dirType: GomokuDirectionType, step: number): number | -1

  /**
   * Move exactly one step in the given direction.
   *
   * If the position after the move is out of bounds, return -1.
   *
   * @param posId     Started position. !!!Should be a valid pos id in bounds.
   * @param dirType   Moving direction.
   */
  safeMoveOneStep(posId: number, dirType: GomokuDirectionType): number | -1

  /**
   * Move the specified number of steps in the given direction.
   *
   * !!! The position after the move should be not out of bounds.
   *
   * @param posId     Started position. !!!Should be a valid pos id in bounds.
   * @param dirType   Moving direction.
   * @param step      Number of steps to move. !!! Should be a non-negative integer.
   */
  fastMove(posId: number, dirType: GomokuDirectionType, step: number): number

  /**
   * Move exactly one step in the given direction.
   *
   * !!! The position after the move should be not out of bounds.
   *
   * @param posId     Started position. !!!Should be a valid pos id in bounds.
   * @param dirType   Moving direction.
   */
  fastMoveOneStep(posId: number, dirType: GomokuDirectionType): number | -1

  /**
   * The maximum number of steps in the given direction can be moved.
   * @param posId
   * @param dirType
   */
  maxMovableSteps(posId: number, dirType: GomokuDirectionType): number

  /**
   * All accessible neighbor positions.
   */
  accessibleNeighbors(posId: number): Iterable<INeighbor>

  /**
   *
   * @param posId
   */
  hasPlacedNeighbors(posId: number): boolean

  /**
   * Get the id of first position in the given direction.
   * @param posId
   * @param dirType
   */
  getStartPosId(posId: number, dirType: GomokuDirectionType): number

  /**
   *
   * @param dirType
   */
  getStartPosSet(dirType: GomokuDirectionType): Iterable<number>

  /**
   *
   * @param handle
   */
  traverseAllDirections(handle: (dirType: GomokuDirectionType) => (posId: number) => void): void
}
