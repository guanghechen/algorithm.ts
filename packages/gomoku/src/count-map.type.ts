import type { GomokuDirectionType } from './constant'
import type { IDirCounter } from './types'

export interface IGomokuCountMap {
  /**
   * Initialize context with given pieces.
   */
  init(): void

  /**
   * Place a piece on the given position.
   * @param posId
   */
  forward(posId: number): void

  /**
   * Remove the piece from the given position.
   * @param posId
   */
  revert(posId: number): void

  /**
   *
   */
  getDirCounters(startPosId: number, dirType: GomokuDirectionType): ReadonlyArray<IDirCounter>
}
