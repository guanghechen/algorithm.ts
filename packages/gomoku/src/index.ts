export * from './constant'
export * from './GomokuContext'
export * from './GomokuCountMap'
export * from './GomokuSolution'
export * from './GomokuState'
export * from './GomokuStateCompressor'
export * from './types'
export * from './util'
