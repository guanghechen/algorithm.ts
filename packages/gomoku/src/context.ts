import { GomokuDirectionTypes, GomokuDirections } from './constant'
import type { GomokuDirectionType } from './constant'
import type { IGomokuContext } from './context.type'
import type { IDirCounter, IGomokuBoard, IGomokuPiece } from './types'
import { createHighDimensionArray } from './util/createHighDimensionArray'

const { full: fullDirectionTypes, rightHalf: halfDirectionTypes } = GomokuDirectionTypes

type IGomokuDirections = number[]
type IIdxMap = Array<Readonly<[r: number, c: number]>>

export interface IGomokuContextProps {
  MAX_ROW: number
  MAX_COL: number
  MAX_ADJACENT: number
  MAX_DISTANCE_OF_NEIGHBOR: number
}

export class GomokuContext implements IGomokuContext {
  public readonly MAX_ROW: number
  public readonly MAX_COL: number
  public readonly MAX_ADJACENT: number
  public readonly MAX_DISTANCE_OF_NEIGHBOR: number
  public readonly TOTAL_PLAYER: number
  public readonly TOTAL_POS: number
  public readonly MIDDLE_POS: number
  public readonly board: Readonly<IGomokuBoard>
  public readonly idx: (r: number, c: number) => number
  protected readonly _idxMap: Readonly<IIdxMap>
  protected readonly _gomokuDirections: Readonly<IGomokuDirections>
  protected readonly _maxMovableMap: number[][] // [dirType][posId] => MAX_MOVABLE_STEPS
  protected readonly _dirStartPosMap: number[][] // [dirType][posId] => startPosId
  protected readonly _dirStartPosSet: number[][] // [dirType] => <start posIds>
  protected readonly _dirNeighborSet: number[][] // [posId] => <neighbors of posId>
  protected readonly _neighborPlacedCount: number[]
  protected readonly _rightHalfDirCountMap: IDirCounter[][][] // [dirType][startPosId] => <Counters>
  protected _placedCount: number

  constructor(props: IGomokuContextProps) {
    const { MAX_ROW, MAX_COL, MAX_ADJACENT, MAX_DISTANCE_OF_NEIGHBOR } = props
    const _MAX_ROW: number = Math.max(1, MAX_ROW)
    const _MAX_COL: number = Math.max(1, MAX_COL)
    const _MAX_ADJACENT: number = Math.max(1, MAX_ADJACENT)
    const _MAX_DISTANCE_OF_NEIGHBOR: number = Math.max(1, MAX_DISTANCE_OF_NEIGHBOR)
    const _TOTAL_PLAYER = 2
    const _TOTAL_POS: number = _MAX_ROW * _MAX_COL
    const idx = (r: number, c: number): number => r * _MAX_ROW + c

    this.MAX_ROW = _MAX_ROW
    this.MAX_COL = _MAX_COL
    this.MAX_ADJACENT = _MAX_ADJACENT
    this.MAX_DISTANCE_OF_NEIGHBOR = _MAX_DISTANCE_OF_NEIGHBOR
    this.TOTAL_PLAYER = _TOTAL_PLAYER
    this.TOTAL_POS = _TOTAL_POS
    this.MIDDLE_POS = _TOTAL_POS >> 1
    this.board = new Array(_TOTAL_POS).fill(-1)
    this.idx = idx

    const _idxMap: IIdxMap = new Array(_TOTAL_POS)
    for (let r = 0; r < _MAX_ROW; ++r) {
      for (let c = 0; c < _MAX_COL; ++c) {
        const posId: number = idx(r, c)
        _idxMap[posId] = [r, c]
      }
    }

    const _gomokuDirections: IGomokuDirections = GomokuDirections.map(
      ([dr, dc]) => dr * MAX_ROW + dc,
    )
    const _maxMovableMap: number[][] = createHighDimensionArray(
      () => 0,
      fullDirectionTypes.length,
      _TOTAL_POS,
    )
    const _dirStartPosSet: number[][] = createHighDimensionArray(
      () => [],
      fullDirectionTypes.length,
    )
    const _dirStartPosMap: number[][] = createHighDimensionArray(
      () => 0,
      fullDirectionTypes.length,
      _TOTAL_POS,
    )
    this.traverseAllDirections(dirType => {
      const revDirType: GomokuDirectionType = dirType ^ 1
      const [dr, dc] = GomokuDirections[dirType]
      return posId => {
        const [r, c] = _idxMap[posId]
        const r2 = r + dr
        const c2 = c + dc
        if (r2 < 0 || r2 >= _MAX_ROW || c2 < 0 || c2 >= _MAX_ROW) {
          _maxMovableMap[dirType][posId] = 0
          _dirStartPosMap[revDirType][posId] = posId
          _dirStartPosSet[revDirType].push(posId)
        } else {
          const posId2 = idx(r2, c2)
          _maxMovableMap[dirType][posId] = _maxMovableMap[dirType][posId2] + 1
          _dirStartPosMap[revDirType][posId] = _dirStartPosMap[revDirType][posId2]
        }
      }
    })

    const _dirNeighborSet: number[][] = new Array(_TOTAL_POS)
    for (let posId = 0; posId < _TOTAL_POS; ++posId) {
      const neighbors: number[] = []
      _dirNeighborSet[posId] = neighbors
      for (const dirType of fullDirectionTypes) {
        let posId2: number = posId
        for (let step = 0; step < _MAX_DISTANCE_OF_NEIGHBOR; ++step) {
          if (1 > _maxMovableMap[dirType][posId2]) break
          posId2 += _gomokuDirections[dirType]
          neighbors.push(posId2)
        }
      }
    }

    const _rightHalfDirCountMap = createHighDimensionArray(
      () => [],
      fullDirectionTypes.length,
      _TOTAL_POS,
    )

    this._gomokuDirections = _gomokuDirections
    this._idxMap = _idxMap
    this._maxMovableMap = _maxMovableMap
    this._dirStartPosMap = _dirStartPosMap
    this._dirStartPosSet = _dirStartPosSet
    this._dirNeighborSet = _dirNeighborSet
    this._neighborPlacedCount = new Array(_TOTAL_POS).fill(0)
    this._rightHalfDirCountMap = _rightHalfDirCountMap
    this._placedCount = 0
  }

  public get placedCount(): number {
    return this._placedCount
  }

  public init(pieces: ReadonlyArray<IGomokuPiece>): void {
    const board = this.board as IGomokuBoard
    board.fill(-1)
    this._placedCount = 0

    // Update _neighborPlacedCount
    const { _neighborPlacedCount } = this
    _neighborPlacedCount.fill(0)
    for (const piece of pieces) {
      const posId: number = this.idx(piece.r, piece.c)
      if (board[posId] >= 0) continue

      board[posId] = piece.p
      this._placedCount += 1
      for (const neighborId of this.accessibleNeighbors(posId)) {
        _neighborPlacedCount[neighborId] += 1
      }
    }

    // update _rightHalfDirCountMap
    const { _rightHalfDirCountMap } = this
    for (const dirType of halfDirectionTypes) {
      for (const startPosId of this.getStartPosSet(dirType)) {
        const counters = _rightHalfDirCountMap[dirType][startPosId]

        let index = 0
        const maxSteps: number = this.maxMovableSteps(startPosId, dirType) + 1
        for (
          let i = 0, posId = startPosId, i2: number, posId2: number;
          i < maxSteps;
          i = i2, posId = posId2
        ) {
          const playerId: number = board[posId]
          for (i2 = i + 1, posId2 = posId; i2 < maxSteps; ++i2) {
            posId2 = this.fastMoveOneStep(posId2, dirType)
            if (board[posId2] !== playerId) break
          }
          // eslint-disable-next-line no-plusplus
          counters[index++] = { playerId, count: i2 - i }
        }
        counters.length = index
      }
    }
  }

  public forward(posId: number, playerId: number): boolean {
    const board = this.board as IGomokuBoard
    if (posId < 0 || posId >= this.TOTAL_POS || board[posId] >= 0) return false

    board[posId] = playerId
    this._placedCount += 1

    // Update _neighborPlacedCount
    for (const neighborId of this.accessibleNeighbors(posId)) {
      this._neighborPlacedCount[neighborId] += 1
    }

    // update _rightHalfDirCountMap
    for (const dirType of halfDirectionTypes) {
      this._updateHalfDirCounter(playerId, posId, dirType)
    }
    return true
  }

  public revert(posId: number): boolean {
    const board = this.board as IGomokuBoard
    if (posId < 0 || posId >= this.TOTAL_POS || board[posId] < 0) return false

    board[posId] = -1
    this._placedCount -= 1

    // Update _neighborPlacedCount
    for (const neighborId of this.accessibleNeighbors(posId)) {
      this._neighborPlacedCount[neighborId] -= 1
    }

    // update _rightHalfDirCountMap
    for (const dirType of halfDirectionTypes) {
      this._updateHalfDirCounter(-1, posId, dirType)
    }
    return true
  }

  public revIdx(posId: number): Readonly<[r: number, c: number]> {
    return this._idxMap[posId]
  }

  public isValidPos(r: number, c: number): boolean {
    return r >= 0 && r < this.MAX_ROW && c >= 0 && c < this.MAX_COL
  }

  public safeMove(posId: number, dirType: GomokuDirectionType, step: number): number {
    return step <= this._maxMovableMap[dirType][posId]
      ? posId + this._gomokuDirections[dirType] * step
      : -1
  }

  public safeMoveOneStep(posId: number, dirType: GomokuDirectionType): number {
    return 1 <= this._maxMovableMap[dirType][posId] ? posId + this._gomokuDirections[dirType] : -1
  }

  public fastMove(posId: number, dirType: GomokuDirectionType, step: number): number {
    return posId + this._gomokuDirections[dirType] * step
  }

  public fastMoveOneStep(posId: number, dirType: GomokuDirectionType): number {
    return posId + this._gomokuDirections[dirType]
  }

  public maxMovableSteps(posId: number, dirType: GomokuDirectionType): number {
    return this._maxMovableMap[dirType][posId]
  }

  public accessibleNeighbors(posId: number): Iterable<number> {
    return this._dirNeighborSet[posId]
  }

  public hasPlacedNeighbors(posId: number): boolean {
    return this._neighborPlacedCount[posId] > 0
  }

  public getStartPosId(posId: number, dirType: GomokuDirectionType): number {
    return this._dirStartPosMap[dirType][posId]
  }

  public getStartPosSet(dirType: GomokuDirectionType): Iterable<number> {
    return this._dirStartPosSet[dirType]
  }

  public getDirCounters(
    startPosId: number,
    dirType: GomokuDirectionType,
  ): ReadonlyArray<IDirCounter> {
    return this._rightHalfDirCountMap[dirType][startPosId]
  }

  public traverseAllDirections(
    handle: (dirType: GomokuDirectionType) => (posId: number) => void,
  ): void {
    const { TOTAL_POS } = this
    const { leftHalf, rightHalf } = GomokuDirectionTypes
    for (const dirType of leftHalf) {
      const h = handle(dirType)
      for (let posId = 0; posId < TOTAL_POS; ++posId) h(posId)
    }
    for (const dirType of rightHalf) {
      const h = handle(dirType)
      for (let posId = TOTAL_POS - 1; posId >= 0; --posId) h(posId)
    }
  }

  protected _updateHalfDirCounter(
    playerId: number,
    posId: number,
    dirType: GomokuDirectionType,
  ): void {
    const revDirType: GomokuDirectionType = dirType ^ 1
    const startPosId = this.getStartPosId(posId, dirType)
    const counters = this._rightHalfDirCountMap[dirType][startPosId]

    let index = 0
    let remain: number = this.maxMovableSteps(posId, revDirType) + 1
    for (; index < counters.length; ++index) {
      const cnt: number = counters[index].count
      if (cnt >= remain) break
      remain -= cnt
    }

    if (remain === 1) {
      if (counters[index].count === 1) {
        const isSameWithLeft: boolean = index > 0 && counters[index - 1].playerId === playerId
        const isSameWithRight: boolean =
          index + 1 < counters.length && counters[index + 1].playerId === playerId
        if (isSameWithLeft) {
          if (isSameWithRight) {
            counters[index - 1].count += 1 + counters[index + 1].count
            counters.splice(index, 2)
          } else {
            counters[index - 1].count += 1
            counters.splice(index, 1)
          }
        } else {
          if (isSameWithRight) {
            counters[index + 1].count += 1
            counters.splice(index, 1)
          } else {
            counters[index].playerId = playerId
          }
        }
      } else {
        counters[index].count -= 1
        if (index > 0 && counters[index - 1].playerId === playerId) {
          counters[index - 1].count += 1
        } else {
          counters.splice(index, 0, { playerId, count: 1 })
        }
      }
    } else if (remain < counters[index].count) {
      const { playerId: hitPlayerId, count: hitCount } = counters[index]
      counters.splice(
        index,
        1,
        { playerId: hitPlayerId, count: remain - 1 },
        { playerId, count: 1 },
        { playerId: hitPlayerId, count: hitCount - remain },
      )
    } else {
      counters[index].count -= 1
      if (index + 1 < counters.length && counters[index + 1].playerId === playerId) {
        counters[index + 1].count += 1
      } else {
        counters.splice(index + 1, 0, { playerId, count: 1 })
      }
    }
  }
}
