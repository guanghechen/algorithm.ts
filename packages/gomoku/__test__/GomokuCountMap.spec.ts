import type { IGomokuPiece } from '../src'
import { GomokuContext, GomokuCountMap, GomokuDirectionType, gomokuDirectionTypes } from '../src'

class TesterHelper {
  public readonly context: GomokuContext
  public readonly board: Int32Array
  public readonly countMap: GomokuCountMap
  protected readonly pieces: IGomokuPiece[]

  constructor(MAX_ROW: number, MAX_COL: number, MAX_INLINE: number) {
    const context = new GomokuContext(MAX_ROW, MAX_COL, MAX_INLINE)
    const board = new Int32Array(context.TOTAL_POS).fill(-1)
    const countMap = new GomokuCountMap(context, board)

    this.context = context
    this.board = board
    this.countMap = countMap
    this.pieces = []
  }

  public init(): void {
    this.board.fill(-1)
    this.countMap.init()
  }

  public forward(r: number, c: number, p: number): void {
    const { context, board, countMap } = this
    const id: number = context.idx(r, c)
    if (board[id] >= 0) return

    countMap.beforeForward(r, c, p)
    board[id] = p
    countMap.afterForward(r, c, p)
  }

  public rollback(r: number, c: number): void {
    const { context, board, countMap } = this
    const id: number = context.idx(r, c)
    if (board[id] < 0) return

    const p: number = board[id]
    countMap.beforeRollback(r, c, p)
    board[id] = -1
    countMap.afterRollback(r, c, p)
  }

  public snapshot(): ReturnType<GomokuCountMap['toJSON']> {
    const board = new Int32Array(this.board)
    for (const { r, c, p } of this.pieces) {
      const id: number = this.context.idx(r, c)
      board[id] = p
    }

    const countMap = new GomokuCountMap(this.context, board)
    countMap.init()
    return countMap.toJSON()
  }
}

describe('7x7', () => {
  const helper = new TesterHelper(7, 7, 5)

  test('basic', function () {
    helper.init()
    const { context, countMap } = helper
    const id1: number = context.idx(1, 1)
    const id2: number = context.idx(2, 2)

    const testState1 = (): void => {
      const { continuouslyShapeCountMap, dirCountMap } = countMap.toJSON()
      expect(continuouslyShapeCountMap[1][1][2]).toEqual(4)
      expect(gomokuDirectionTypes.every(dirType => dirCountMap[dirType][id1] === 1)).toEqual(true)
    }

    const testState2 = (): void => {
      const { continuouslyShapeCountMap, dirCountMap } = countMap.toJSON()
      expect(continuouslyShapeCountMap[1][1][2]).toEqual(6)
      expect(continuouslyShapeCountMap[1][2][2]).toEqual(1)
      expect(
        gomokuDirectionTypes.every(
          dirType =>
            dirCountMap[dirType][id1] === (dirType === GomokuDirectionType.BOTTOM_RIGHT ? 2 : 1),
        ),
      ).toEqual(true)
      expect(
        gomokuDirectionTypes.every(
          dirType =>
            dirCountMap[dirType][id2] ===
            ((dirType ^ 1) === GomokuDirectionType.BOTTOM_RIGHT ? 2 : 1),
        ),
      ).toEqual(true)
    }

    helper.forward(1, 1, 1)
    testState1()

    helper.forward(2, 2, 1)
    testState2()

    helper.forward(3, 3, 1)
    helper.forward(4, 4, 1)
    helper.forward(5, 5, 1)
    helper.forward(5, 2, 1)
    helper.forward(4, 2, 0)
    helper.forward(4, 0, 0)
    expect(countMap.toJSON()).toMatchSnapshot()

    helper.rollback(4, 0)
    helper.rollback(4, 2)
    helper.rollback(5, 2)
    helper.rollback(5, 5)
    helper.rollback(4, 4)
    helper.rollback(3, 3)
    testState2()
    helper.rollback(2, 2)
    testState1()
  })

  test('complicate', function () {
    helper.init()
    expect(helper.countMap.toJSON()).toEqual(helper.snapshot())
    helper.forward(1, 1, 1)
    expect(helper.countMap.toJSON()).toEqual(helper.snapshot())
    helper.forward(2, 3, 1)
    expect(helper.countMap.toJSON()).toEqual(helper.snapshot())
    helper.forward(1, 3, 0)
    expect(helper.countMap.toJSON()).toEqual(helper.snapshot())
    helper.forward(3, 1, 1)
    expect(helper.countMap.toJSON()).toEqual(helper.snapshot())
    helper.forward(4, 1, 0)
    expect(helper.countMap.toJSON()).toEqual(helper.snapshot())
    helper.forward(5, 3, 1)
    expect(helper.countMap.toJSON()).toEqual(helper.snapshot())
    helper.forward(6, 4, 1)
    expect(helper.countMap.toJSON()).toEqual(helper.snapshot())
    helper.rollback(5, 3)
    expect(helper.countMap.toJSON()).toEqual(helper.snapshot())
    helper.forward(1, 1, 1)
    expect(helper.countMap.toJSON()).toEqual(helper.snapshot())
    helper.rollback(4, 1)
    expect(helper.countMap.toJSON()).toEqual(helper.snapshot())
    helper.forward(3, 6, 1)
    expect(helper.countMap.toJSON()).toEqual(helper.snapshot())
    helper.rollback(1, 1)
    expect(helper.countMap.toJSON()).toEqual(helper.snapshot())
    helper.forward(0, 0, 1)
    expect(helper.countMap.toJSON()).toEqual(helper.snapshot())
    helper.forward(0, 1, 1)
    expect(helper.countMap.toJSON()).toEqual(helper.snapshot())
    helper.forward(0, 3, 1)
    expect(helper.countMap.toJSON()).toEqual(helper.snapshot())
    helper.forward(0, 4, 1)
    expect(helper.countMap.toJSON()).toEqual(helper.snapshot())
    helper.forward(0, 5, 0)
    expect(helper.countMap.toJSON()).toEqual(helper.snapshot())
  })
})
