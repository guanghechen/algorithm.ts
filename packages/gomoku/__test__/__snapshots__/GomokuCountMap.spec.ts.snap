// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`7x7 basic 1`] = `
Object {
  "continuouslyShapeCountMap": Array [
    Array [
      Array [
        0,
        0,
        0,
      ],
      Array [
        0,
        5,
        3,
      ],
      Array [
        0,
        0,
        0,
      ],
      Array [
        0,
        0,
        0,
      ],
      Array [
        0,
        0,
        0,
      ],
      Array [
        0,
        0,
        0,
      ],
      Array [
        0,
        0,
        0,
      ],
      Array [
        0,
        0,
        0,
      ],
      Array [
        0,
        0,
        0,
      ],
      Array [
        0,
        0,
        0,
      ],
    ],
    Array [
      Array [
        0,
        0,
        0,
      ],
      Array [
        0,
        2,
        17,
      ],
      Array [
        0,
        0,
        0,
      ],
      Array [
        0,
        0,
        0,
      ],
      Array [
        0,
        0,
        0,
      ],
      Array [
        0,
        0,
        1,
      ],
      Array [
        0,
        0,
        0,
      ],
      Array [
        0,
        0,
        0,
      ],
      Array [
        0,
        0,
        0,
      ],
      Array [
        0,
        0,
        0,
      ],
    ],
  ],
  "dirCountMap": Array [
    Uint32Array [
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      1,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      1,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      1,
      0,
      0,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      0,
      0,
      0,
      1,
      0,
      0,
      1,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
    ],
    Uint32Array [
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      1,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      1,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      1,
      0,
      0,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      0,
      0,
      0,
      1,
      0,
      0,
      1,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
    ],
    Uint32Array [
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      1,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      2,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      3,
      0,
      0,
      0,
      1,
      0,
      1,
      0,
      4,
      0,
      0,
      0,
      0,
      1,
      0,
      0,
      5,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
    ],
    Uint32Array [
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      5,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      4,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      3,
      0,
      0,
      0,
      1,
      0,
      1,
      0,
      2,
      0,
      0,
      0,
      0,
      1,
      0,
      0,
      1,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
    ],
    Uint32Array [
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      1,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      1,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      1,
      0,
      0,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      0,
      0,
      0,
      1,
      0,
      0,
      1,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
    ],
    Uint32Array [
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      1,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      1,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      1,
      0,
      0,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      0,
      0,
      0,
      1,
      0,
      0,
      1,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
    ],
    Uint32Array [
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      1,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      1,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      1,
      0,
      0,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      0,
      0,
      0,
      1,
      0,
      0,
      1,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
    ],
    Uint32Array [
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      1,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      1,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      1,
      0,
      0,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      0,
      0,
      0,
      1,
      0,
      0,
      1,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
    ],
  ],
  "gapShapeCountMap": Array [
    Array [
      Array [
        0,
        0,
        0,
      ],
      Array [
        0,
        0,
        0,
      ],
      Array [
        0,
        1,
        0,
      ],
      Array [
        0,
        0,
        0,
      ],
      Array [
        0,
        0,
        0,
      ],
      Array [
        0,
        0,
        0,
      ],
      Array [
        0,
        0,
        0,
      ],
      Array [
        0,
        0,
        0,
      ],
      Array [
        0,
        0,
        0,
      ],
      Array [
        0,
        0,
        0,
      ],
    ],
    Array [
      Array [
        0,
        0,
        0,
      ],
      Array [
        0,
        0,
        0,
      ],
      Array [
        0,
        0,
        0,
      ],
      Array [
        0,
        0,
        0,
      ],
      Array [
        0,
        0,
        0,
      ],
      Array [
        0,
        0,
        0,
      ],
      Array [
        0,
        0,
        0,
      ],
      Array [
        0,
        0,
        0,
      ],
      Array [
        0,
        0,
        0,
      ],
      Array [
        0,
        0,
        0,
      ],
    ],
  ],
}
`;
